js/interface.js

```diff
@@ -315,14 +311,103 @@
+  function isNewerVersion(latestVersionMenu, currentMenuVersion) {
+    for (var index = 0; index < latestVersionMenu.length; index++) {
+      if (isNaN(+latestVersionMenu[index])) latestVersionMenu[index] = 0;
+
+      if (isNaN(+currentMenuVersion[index])) currentMenuVersion[index] = 0;
+
+      return +latestVersionMenu[index] > +currentMenuVersion[index];
+    }
+
+    return false;
+  }
+
+  /**
+   * Function to filter organizationMenus by version.
+   * @param { Array } data - array of organizationMenus
+   * @returns { Object } latestVersionMenu - object which contains the latest version of current menu package type
+   */
+
+  function getLatestMenuVersion(data) {
+    var latestVersionMenu;
+    var previousMenu;
+
+    data.forEach(function(menu) {
+      var formattedMenuVersion = menu.version.split('.');
+      var formattedLatestVersionMenu = previousMenu ? previousMenu.version.split('.') : null;
+
+      if (previousMenu) {
+        latestVersionMenu = isNewerVersion(formattedLatestVersionMenu, formattedMenuVersion) ? previousMenu : menu;
+      } else {
+        latestVersionMenu = menu;
+      }
+
+      previousMenu = menu;
+    });
+
+    return latestVersionMenu;
+  }
+
+  /**
+   * Creates a list of menu widgets to be displayed based on versions and current usage
+   * @param { Array } menus - List of menu widgets
+   * @returns { Array } List of menu widgets to be displayed based on versions and current usage
+   */
+
+  function generateMenuList(menus) {
+    var menusByPackage = _.groupBy(menus, 'package');
+    var menuList = [];
 
+    _.forIn(menusByPackage, function(menuVersions) {
+      var currentVersion = _.find(menuVersions, function(menuVersions) {
+        return menuVersions.instances.length;
+      });
+
+      // A version is currently in use
+      if (currentVersion) {
+        return menuList.push(currentVersion);
+      }
+
+      // Find latest version of menu
+      menuList.push(getLatestMenuVersion(menuVersions));
+    });
+
+    // Sort displayed menus by display name
+    return _.sortBy(menuList, function(menu) {
+      return menu.name.trim().toUpperCase();
+    });
+  }
+
+  function loadCustomMenuWidgets() {
     return fetchCustomMenuWidgets().then(function(menus) {
-      customMenus = menus;
+      var menusWithInstances = _.filter(menus, function(menu) {
+        return menu.instances.length;
+      });
+
+      // If there is more than 1 menu with instances, clean up
+      if (menusWithInstances.length > 1) {
+        // Keep the first one found
+        menusWithInstances.shift();
+
+        var instancesToDelete = _.flatten(_.map(menusWithInstances, function(menu) {
+          return _.map(menu.instances, 'id');
+        }));
+
+        // Delete unneeded instances and fetch menus again
+        return Promise.all(instancesToDelete.map(function(item) {
+          return deleteInstance({ id: item });
+        })).then(fetchCustomMenuWidgets());
```

Comment:
I don't think you have tested this `.then()` condition, because `.then()` takes a `Function` as parameter, not a `Promise`.

```suggestion
        })).then(fetchCustomMenuWidgets);
```
```suggestion
        })).then(fetchCustomMenuWidgets());
```

Reply:
@tonytlwu hi, now tested and fixed